// react-native-expo-image-editor/lib/commonjs/components/Grid/GridEdge.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = GridEdge;

var _react = _interopRequireDefault(require("react"));

var _reactNative = require("react-native");

var _Grid = _interopRequireDefault(require("../../styles/Grid.style"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// src/compoennts/Grid/GridEdge.tsx

/**
 *
 *
 * @export
 * @param {IEdgeProps} props
 * @return {*}  {React.ReactElement}
 */
function GridEdge(props) {
  const {
    left,
    top,
    bottom,
    right,
    alignSelf
  } = props; // console.log("Edge reload")

  const getStyles = () => {
    return {
      borderLeftWidth: left ? left : 0,
      borderTopWidth: top ? top : 0,
      borderBottomWidth: bottom ? bottom : 0,
      borderRightWidth: right ? right : 0,
      alignSelf
    };
  };

  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [getStyles(), _Grid.default.edge]
  });
}
