// react-native-expo-image-editor/lib/commonjs/components/Controller.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Controller;

var _react = _interopRequireDefault(require("react"));

var _reactNative = require("react-native");

var _Provider = require("./Provider");

var _Row = _interopRequireDefault(require("./Row"));

var _Label = _interopRequireDefault(require("./Label"));

var _Controller = _interopRequireDefault(require("../styles/Controller.style"));

var _Button = _interopRequireDefault(require("./Button"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// src/components/Controller.tsx
function Controller(props) {
  const {
    children,
    onBack,
    onDone
  } = props;
  const {
    manipulator: {
      setVisible
    }
  } = (0, _Provider.useProvider)();

  const handleCrop = () => {
    setVisible(true);
  };

  return /*#__PURE__*/_react.default.createElement(_reactNative.SafeAreaView, {
    style: _Controller.default.container
  }, /*#__PURE__*/_react.default.createElement(_Row.default, null, /*#__PURE__*/_react.default.createElement(_Label.default, {
    onPress: onBack
  }, "\u9000\u51FA"), /*#__PURE__*/_react.default.createElement(_Label.default, {
    onPress: handleCrop
  }, "\u526A\u88C1")), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: {
      alignItems: 'center',
      justifyContent: 'center'
    }
  }, children), /*#__PURE__*/_react.default.createElement(_Row.default, null, /*#__PURE__*/_react.default.createElement(_Label.default, null), /*#__PURE__*/_react.default.createElement(_Button.default, {
    onPress: onDone
  }, "\u5B8C\u6210")));
}
