// react-native-expo-image-editor/lib/commonjs/components/FadeView.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactNative = require("react-native");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class FadeView extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      fadeAnim: new _reactNative.Animated.Value(0)
    });
  }

  componentDidUpdate() {
    _reactNative.Animated.timing(this.state.fadeAnim, {
      toValue: this.props.visible ? 1 : 0,
      duration: 300,
      useNativeDriver: true
    }).start();
  }

  componentWillUnmount() {
    _reactNative.Animated.timing(this.state.fadeAnim, {
      toValue: 0,
      duration: 300,
      useNativeDriver: true
    }).start();
  }

  render() {
    const {
      visible,
      children,
      style
    } = this.props;

    const getFadeStyle = () => ({
      opacity: this.state.fadeAnim
    });

    return /*#__PURE__*/_react.default.createElement(_reactNative.Animated.View, {
      pointerEvents: visible ? 'auto' : 'none',
      style: [getFadeStyle(), style]
    }, children);
  }

}

exports.default = FadeView;
