// react-native-expo-image-editor/lib/commonjs/components/Provider.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useProvider = exports.Provider = exports.EditorContext = exports.Consumer = void 0;

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// src/components/Provider.tsx
const initialState = {
  imageUri: '',
  setImageUri: () => {},
  manipulator: {
    visible: false,
    label: 'Rotate',
    setVisible: () => {}
  }
};

const EditorContext = /*#__PURE__*/_react.default.createContext(initialState);

exports.EditorContext = EditorContext;

const Provider = props => {
  const {
    uri,
    children
  } = props;

  const [imageUri, setImageUri] = _react.default.useState(uri);

  const [visible, setVisible] = _react.default.useState(false);

  return /*#__PURE__*/_react.default.createElement(EditorContext.Provider, {
    value: {
      imageUri,
      setImageUri,
      manipulator: {
        visible,
        setVisible,
        label: ''
      }
    }
  }, children);
};

exports.Provider = Provider;
const Consumer = EditorContext.Consumer;
exports.Consumer = Consumer;

const useProvider = () => _react.default.useContext(EditorContext);

exports.useProvider = useProvider;
