// react-native-expo-image-editor/lib/module/components/Controller.js
// src/components/Controller.tsx
import React from 'react';
import { View, SafeAreaView } from 'react-native';
import { useProvider } from './Provider';
import Row from './Row';
import Label from './Label';
import styles from '../styles/Controller.style';
import Button from './Button';
export default function Controller(props) {
  const {
    children,
    onBack,
    onDone
  } = props;
  const {
    manipulator: {
      setVisible
    }
  } = useProvider();

  const handleCrop = () => {
    setVisible(true);
  };

  return /*#__PURE__*/React.createElement(SafeAreaView, {
    style: styles.container
  }, /*#__PURE__*/React.createElement(Row, null, /*#__PURE__*/React.createElement(Label, {
    onPress: onBack
  }, "\u9000\u51FA"), /*#__PURE__*/React.createElement(Label, {
    onPress: handleCrop
  }, "\u526A\u88C1")), /*#__PURE__*/React.createElement(View, {
    style: {
      alignItems: 'center',
      justifyContent: 'center'
    }
  }, children), /*#__PURE__*/React.createElement(Row, null, /*#__PURE__*/React.createElement(Label, null), /*#__PURE__*/React.createElement(Button, {
    onPress: onDone
  }, "\u5B8C\u6210")));
}
