// react-native-expo-image-editor/lib/module/components/FadeView.js
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// src/components/FadeView.tsx
import React from 'react';
import { Animated } from 'react-native';
export default class FadeView extends React.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      fadeAnim: new Animated.Value(0)
    });
  }

  componentDidUpdate() {
    Animated.timing(this.state.fadeAnim, {
      toValue: this.props.visible ? 1 : 0,
      duration: 300,
      useNativeDriver: true
    }).start();
  }

  componentWillUnmount() {
    Animated.timing(this.state.fadeAnim, {
      toValue: 0,
      duration: 300,
      useNativeDriver: true
    }).start();
  }

  render() {
    const {
      visible,
      children,
      style
    } = this.props;

    const getFadeStyle = () => ({
      opacity: this.state.fadeAnim
    });

    return /*#__PURE__*/React.createElement(Animated.View, {
      pointerEvents: visible ? 'auto' : 'none',
      style: [getFadeStyle(), style]
    }, children);
  }

}
